emscripten-simulated-jerry-api/jerry.js

```diff
@@ -0,0 +1,306 @@
+            writable: false,
+            value: internalProps,
+          });
+        }
+      }
+      var entry = {
+        jerry_value: jerry_value,
+        refCount: 1,
+        value: value,
+        error: false,
+        internalProps: internalProps
+      };
+      // Install garbage collection callback:
+      this.weak(value, _jerryhandleGarbageCollected.bind(this, internalProps));
+      this._jerryToHostValueMap[jerry_value] = entry;
+      // console.log('created entry ' + jerry_value + ' for ' + value + ' at ' + stackTrace());
+      return jerry_value;
+    },
+
+    reset: function () {
+      this._jerryToHostValueMap = {};
+      this._nextObjectRef = 1;
+    },
+
+    // Given a jerry value, return the stored javascript value.
+    get: function (jerry_value) {
+      return this._getEntry(jerry_value).value;
+    },
+
+    // Given a javascript value, return a jerry value that refers to it.
+    // If the value already exists in the map, increment its refcount and return
+    // the jerry value.
+    // Otherwise, create a new entry and return the jerry value.
+    ref: function (value) {
+      var jerry_value;
+      var entry;
+      // Fast path for already known/marked objects:
+      if (hasInternalProps(value)) {
+        jerry_value = value.__jerryInternalProps.jerry_value;
+        // Note: It's possible there is no entry in the map because the native
+        // side did not retain it any more, we need to register it again in that case.
+        entry = this._jerryToHostValueMap[jerry_value];
+      } else {
+        entry = this._findEntryByHostValue(value);
+      }
+      if (entry) {
+        ++entry.refCount;
+        return jerry_value || entry.jerry_value;
+      }
+      return this._registerHostValue(value, jerry_value);
+    },
+
+    getRefCount: function (jerry_value) {
+      return this._getEntry(jerry_value).refCount;
+    },
+
+    // Increase the reference count of the given jerry value
+    acquire: function (jerry_value) {
+      this._getEntry(jerry_value).refCount++;
+      return jerry_value;
+    },
+
+    // Decrease the reference count of the given jerry value and delete it if
+    // there are no more internal references.
+    release: function (ref) {
+      var entry = this._getEntry(ref);
+      entry.refCount--;
+
+      if (entry.refCount <= 0) {
+        // Don't remove primitives like string, number, etc because
+        // we would otherwise loose the jerry_value that should be
+        // associated with it. (It's not possible to add the
+        // __jerryInternalProps property w/o 'boxing' the primitive).
+        // The value might get 'sent back' to the native side later on
+        // and we want to make sure the jerry_value remains the same for
+        // the entirety of the lifetime of the value.
+        if (hasInternalProps(entry.value)) {
+          // console.log('deleting ' + ref + ' at ' + stackTrace());
+          delete this._jerryToHostValueMap[ref];
```

Comment:
@martijnthe this will lead to a bug:

In native binding code: I save the jerry_value_t in some place (e.g, saved as the callback's args), but keep the refcount of the jerry_value is 0
In js: the js value is accessed by root, so it will not be GCed even if its jerry_value's refcount is 0.

When a event triggers the callback,  the system will call the callback with the saved jerry_value. But in the emx, because the refcount of the jerry_value is 0, the (ref: js_value) pair is deleted from _jerryToHostValueMap, and the 
system cant find the corrsponding js value by the saved jerry_value. 

----------------

And we cant "comment out" the delete statement, cause they will disable all gc of the js value, and may fail the test cases which contains the native free cb.

Or maybe we should "comment out" the delete statement, and disable all gc in emx ? Then we should move all native free cb related test cases into a seperate file, and dont test them in emx-unittest.

Reply:
Ooh, good catch!

The node `weak` module returns a "weak reference" object.
https://www.npmjs.com/package/weak
I think we should use that "weak reference" instead of setting the host object directly as a property of the `entry` objects in  `_jerryToHostValueMap`.

In case the `weak` module isn't available (i.e. browser), I don't think there's a good solution and "just leaking" everything is the only alternative.

I'll work on this tomorrow.